### DISCUSSION TITLE
SZTU 2022暑期集训 记忆化搜索与竞赛技巧

### DISCUSSION THREAD

**Discussion by ColdSnapdragon** [CSGrandeur/s-1problem1day1ac#464] [Algorithm]
# Day 1

## 记忆化搜索

**记忆化搜索=搜索的形式+动态规划的思想**

*经典例子——计算斐波那契数列第x项*

朴素的DFS递归求解：时间复杂度为O(2^n)

```cpp
long long fibo(int x)
{
    if(x==1 || x==2) return 1LL;
    return fibo(x-1)+fibo(x-2);
}//1s内最多只能计算x=40
```

线性递推求解（动态规划）：时间复杂度为O(n)

```cpp
fb[1]=fb[2]=1;
for(int i=3;i<=x;++i)
	fb[i]=fb[i-1]+fb[i-2]; //1s内能计算到10^8
```

造成递归求解的指数级时间复杂度的根本原因，是递归的过程中进行了大量的重复计算。

### 一、记忆化搜索(Mnemonic search)

基本思想：如果我们能记录一些状态的答案，在重新到达该状态时直接获得结果，就可以减少重复搜索量。**记忆化搜索是一种通过记录已经遍历过的状态的信息，从而避免对同一状态重复遍历的搜索实现方式。**

实现方式：在朴素DFS做法上，加入一个数组来记录已经计算过的值。数组元素初始化为0，在fibo(x)计算出来后，记录下其结果，待到下次需要重新计算fibo(x)时，直接返回该值。

记忆化搜索：时间复杂度为O(n)

```cpp
long long fibo(int x)
{//fb[x]初始化为0
    if(x==1 || x==2) return 1LL;
    if(fb[x]!=0) return fb[x];//如果曾计算过fibo(x)，直接返回这个结果
    fb[x] = fibo(x-1)+fibo(x-2);//记录
    return fb[x];
}
```

**记忆化搜索的核心实现：**

参考模板：

```cpp
返回类型 dfs(形参)
{
    if(终止条件)
        return 结果;
    if(本状态有记录)
        return 记录的结果;
    朴素DFS;
    将结果记录至数组;
    return 结果;
}
```

### 二、算法核心

（一）算法细节

1 通常地，终止条件应该放在记忆化搜索之前（以避免细节bug）。

2 记忆数组mem为全局变量，数组中的每个元素必须能代表唯一的状态，其维度一般由参数数量确定，例如：

```cpp
int dfs(int x,int y,int z)
{
	...
	if(mem[x][y][z]!=0)
        return mem[x][y][z];//三维
	...
}
```

在记忆化搜索中，例如上面的代码，绝不能出现例如”先后执行两次dfs(1,2,3)，却产生不同的搜索过程”的情况。**因此，记忆化过程中通常不能引入其他全局变量。**

```cpp
void dfs(int p,int k)
{
    ...
	for(int i=k;i<=n;++i)
    {
        ar[p]=i;
        dfs(p+1,i);
        ar[p]=0;
    }
    ...
} //如上，我们并没有办法记忆化搜索
```

3 如果搜索结果可能为0，那么记忆数组就不能初始化为0，可以将其初始化为-1或其他特殊值。

```cpp
if(~mem[x][y]) //~(-1)=0
	return mem[x][y];
```

4 如果有多组样例，不要用memset初始化记忆数组。

（二）复杂度评价

```cpp
int dfs(int x,int y)
{
    ...
    if(~mem[x][y])
   		return mem[x][y];
    for(int i=x+1;i<=n;++i)
        dfs(i,y+1);
    ...
}
```

记忆化搜索的时间复杂度为：所有状态的总访问次数（不是状态总数）。

大致计算：理论状态数*每个节点的复杂度（在每个结点内跑for就会再产生O(n)复杂度）。

以上代码的时间复杂度为O(n^3)。

### 三、与动态规划的联系

因为记忆化搜索（以下简称mem）确保了每个状态只访问一次，它也是一种常见的动态规划实现方式。

记忆化搜索和动态规划是真包含关系，可相互转化的DP和mem本质都是同一个状态转移方程。DP侧重于从前往后转移，而记忆化搜索则是从后向前归纳，在同等复杂度下，dp的效率会更高，与mem出现显著用时差距，因此，如果能简单写出递推式就不要用记忆化搜索。如果思考DP的转态表达和状态转移遇到困难，也不妨从DFS的角度思考转移方式。

DP和mem在某些情况下不能相互转换：

对于记忆化搜索，当状态转移存在复杂的拓扑关系，或者需要哈希状态时，就难以写出DP。

对于动态规划，当需要用到前缀和优化、斜率优化等时，就不能写成记忆化搜索。

### 四、解答树、DAG

记忆化搜索所基于的解答树，实际上是一个有向无环图（DAG），其最为关键的特征就是“不返祖”（后续的转态无法通过任意方式返回之前转移向自己的状态，即“无后效性”）。**其中，点为状态，有向边为决策方向，从起点到端点的一条路径为一段决策序列**。*记忆化搜索的总复杂度为边的数量*。

![248px-Transitive_reduction_of_G](https://user-images.githubusercontent.com/55649366/180674128-a0a4337d-12e3-43f7-819f-4ba6473c3183.png)

当第二次到达一个点时，就会在之后做出一系列与此前完全相同的决策，我们通过记忆化搜索来储存结果，从而避免了重复的搜索。

如果一个问题能抽象成DAG上的问题，那么就可以考虑采用记忆化搜索来处理。

考虑到状态的转移是单向的，也可以在对DAG图拓扑排序后，从左到右进行动态规划。

![查看源图像](https://tse2-mm.cn.bing.net/th/id/OIP-C.eoErBTRL4pZ4xx0jNF1GrgHaCo?pid=ImgDet&rs=1)

除此之外，对于一些非DAG的有向图而言，将图中的强连通分量缩点，就成为了DAG图，也可以结合问题运用记忆化搜索的思想。

**题单**（待更新）
<html>
<body>
<!DOCTYPE html><figure class="md-table-fig" cid="n51" mdtype="table" style="box-sizing: border-box; margin: 1.2em 0px; overflow-x: auto; max-width: calc(100% + 16px); padding: 0px; cursor: default; color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

题号 | 标签 | 难度
<html>
<body>
<!DOCTYPE html><figure class="md-table-fig" cid="n49" mdtype="table" style="box-sizing: border-box; margin: 1.2em 0px; overflow-x: auto; max-width: calc(100% + 16px); padding: 0px; cursor: default; color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><br class="Apple-interchange-newline">

<html>
<body>
<!DOCTYPE html><br class="Apple-interchange-newline">

题号 | 标签 | 难度
-- | -- | --
POJ 1579 | 记忆化搜索 | ★
计蒜客-T1408 矩形嵌套 | DAG记忆化搜索 | ★★
POJ 1088 滑雪 | DAG记忆化搜索 | ★★
洛谷P3183 | DAG记忆化搜索 | ★★
洛谷P1025 | 记忆化搜索 不降原则 | ★★
Atcoder abc231_e | 离散化+记忆化搜索 | ★★★
洛谷P2476 着色方案 | 记忆化搜索 | ★★★★

</body>
</html>

## 竞赛技巧

![image-20220724192049819](https://user-images.githubusercontent.com/55649366/180674137-dd908bd0-dbaf-4e1d-9d9d-a1b57d847755.png)

我们做题目时都会根据数据范围，预计时间复杂度。如果不幸被卡掉，那是因为我们程序的常数太大。（预计复杂度$O(n^2)$，实际总复杂度$O(10n^2+nlogn+n+...)$）

常见的几种优化方式有更改输入输出方式，优化写法，位运算行各种操作使得程序更快。

### 一、打表

朴素的打表，指的是在比赛时把所有可能的输入对应的答案都计算出来并保存下来，然后在代码里开个数组把答案放里面，直接输出即可。总体思路就是以空间换时间。

```cpp
//多组数据，每次输入一个正整数n(2<=n<=10^5)，输出n以内的质数
```

在想不出更好办法的情况下（PS:我们其实有更好的办法），那么要怎么解决时间超限的问题呢?

我们可以打小抄

```cpp
#include<iostream>
using namespace std;
bool judge(int x)
{
    int j=1;
    for(int i=2;i<x/2+1;i++)
        if(x%i==0)
            j=0;
    return j;
}
int main()
{
	freopen("tem.txt","w",stdout); 
    int n=100000;
    for(int i=2;i<n;i++)
    {   
        if(i%2==0&i!=2)
            continue;
        if(judge(i))
            cout<<i<<",";//这里用逗号分隔开，方便我们一会儿存入数组中
    }
    return 0;
}
```

程序将输出：

```
2,3,5,7,11,13,17,19,23,29,31, ... ,99907,99923,99929,99961,99971,99989,99991,
```

接着：

```cpp
int ar[100000]={/*粘贴到此处*/};
```

之后就可以根据题意任意输出结果了。

**分段打表**

*例题：规定$f(x)$为整数x的二进制表示中1的个数。输入一个正整数n (n<1e9)，输出$\sum_{i=1}^{n}f(i)^2$ 。*

我们可以在O(logn)下处理出一个数字x的f(x)。如果仍用朴素的打表法，我们开不出这么大的数组。

我们考虑优化这个答案表。采用分块思想，我们设置一个合理的步长 （这个步长一般视代码长度而定），对于第i块，计算出这块范围下的$f(x)^2$之和。在输出结果时，整块的答案用预处理的值计算，非整块的答案暴力计算。

可以取每块的长度为1e6，于是我们只需要预先打表处理出1e9/1e6=1000个数，并在nlongn时间内完成本题。

### 二、对拍

`我们有瞪眼法来debug（`

什么是对拍？ 当我们的程序过了样例，是否意味着它一定能AC呢？显然大多数情况下都是不行的。所以我们需要自己设计一些数据来测试我们的程序，但有的题目数据很大，我们肉眼无法看出程序计算的结果是否正确，手工计算又非常耗时，在紧张的比赛中，我们该怎么应对呢？于是有了对拍。 对拍简单的说就是当你写完一个题目的程序以后，再写一个暴力求解该题目的程序，然后自己生成一些测试数据，看同样的数据，两个程序输出的结果是否相同，不同意味着被对拍的程序有问题。以此来帮助你修改程序，提高通过率的方法，我们称为对拍。

对拍由四部分组成：

1. 一个随机数据生成器
2. 你的暴力
3. 你的“疑似正解”
4. 对拍程序

对比暴力结果和自己的代码结果，就可以找到合适的数据来debug。

————————————————————————————————————————————

除了留意简化冗余代码，在必要时，还可以对程序进行**常数优化**。所谓常数优化，就是优化总复杂度中常数项，使它们变得更小。

### 三、输入输出优化

这是最快与最重要的优化方式。

众所周知，对于C++来说有三种常见的读入方式即`cin，scanf，read`（快读），这几种读入方式各有优点和缺点。

`cin`，C++标准的读入方式，读入字符串有得天独厚的优势，适应性强，使用简单，方便，可以文件输入同步，缺点是速度十分慢。
`scanf`，适合格式化输入：擅长精度控制、隔行输入等十分方便，常用于处理一些数字中间掺和符号的读入，但处理字符串没有`cin`方便，但速度比`cin`快很多，功能复杂多样，是C++最常见的读入方式。
`read`，又称读入优化，其优化原理是基于C++`getchar`运算速度非常快，可以实现高效率读入大规模数据，且就有很强的伸缩性，可以在其函数内进行一些特判，例如跳过字母等等，缺点是对新手来说有点难理解，是中高阶选手普遍普遍使用的读入方式，缺点是输入单一，处理数据功能弱

```cpp
inline int read()
{
    char c = getchar();int x = 0,f = 1;
    while(c < '0' || c > '9') {if(c == '-') f = -1;c = getchar();}//符号
    while(c >= '0' && c <= '9') {x = x*10 + c -'0';c = getchar();}//数字
    return x*f;
}
int a=read();

```

关于`cin cout` 和 `scanf printf`，做题的时候尽量使用`scanf printf`。`cin cout` 比 `scanf printf`慢20倍左右！

`cin，cout`之所以效率低，是因为先把要输出的东西存入缓冲区，再输出，导致效率降低。可以使用`std::ios::sync_with_stdio(false);`这条语句关掉`scanf` 和`cin` 的同步，取消`iostream`的输入 输出缓存，从而加快效率。

```cpp
int main()
{
    cin.sync_with_stdio(false); //提升cin,cout的速度
    cin.tie(0),cout.tie(0);
    ...
}
```

*注意，关闭同步后，不能再使用`scanf, printf`，避免出现错误。*

### 四、其他常数优化

位操作具备更快的运算速度

```cpp
x=x*2;
//改为
x<<=1;
    
y=y/2;
//改为
y>>=1;
```

判断奇偶数：

```cpp
if(x%2==1) ;
//改为
if(x&1) ;
```

求2的n次方

```cpp
x=pow(2,n);
//改为
x=(1<<n);
```

register关键字声明，请求在寄存器上分配空间

```cpp
for(int i=1;i<=n;i++){}
//改为
for(register int i=1;i<=n;++i){}
```

bool的速度慢于int

```cpp
bool tag[100000];
//改为
int tag[100000];
```

对非递归且代码块短的函数，声明为内联(inline)

```cpp
int cmp(int a,int b) {return a<b;}
//改为
inline int cmp(int a,int b) {return a<b;}
```

每次调用函数时都会重新创建它的形参，并用传入的实参对形参进行初始化（实参被**值传递**时相当于创建实参的拷贝），拷贝大的类类型对象或者容器对象比较低效（比如vector，或者一个很长的string），如果函数无须改变形参的值，最好将其声明为常量引用。

```cpp
inline int cmp(string a,string b) {return a<b;}
//改为
inline int cmp(const string &a,const string &b) {return a<b;}
```

**参考题单（亟待补充）**
<html>
<body>
<!DOCTYPE html><figure class="md-table-fig" cid="n144" mdtype="table" style="box-sizing: border-box; margin: 1.2em 0px; overflow-x: auto; max-width: calc(100% + 16px); padding: 0px; cursor: default; color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><br class="Apple-interchange-newline">

题号 | 标签 | 难度
-- | -- | --
BZOJ p3798 | 分段打表 | ★★
洛谷 P1822 | 分段打表 | ★★★

</figure><br class="Apple-interchange-newline">
</body>
</html>

**Comment by cyh834** [comment]
常量用const修饰会更快，洛谷P4783 不const int mod会TLE

**Reply by SamuNatsu** [reply]
一般如数字（整数如：`123`、`456ll`、`0x789`；浮点数如：`1.0`、`1e-8`）、字符串（如：`"abc"`）、字符（如：`'a'`）等字面量一般不使用 const 来声明常量。

const 会存储在程序的全局常量区，在实际运算中是通过读取对应位置的内存来获得常量值（特别是常量为一个表达式的情况下），实际上会消耗部分寻址性能。至于 const 能比非 const 快那是因为编译器针对 const 会有特殊的优化，但是这个在不同的编译器上是不同的，并不稳定。

字面量常量应该使用 define 来定义，因为 define 的工作只是将对应的宏改写成字面量，这样编译成机器代码后就会变成直接对字面值进行操作，不需要进行寻址和内存读取，这个没有编译器区别，是稳定的。

但是如果常量是一个表达式，那么还是应该使用 const，因为表达式不是字面量，其需要经过计算，使用 define 的话很容易想到会将表达式展开，反而会加重计算负担。

**Comment by SamuNatsu** [comment]
# 重要指正

## register 关键字问题

**从 C++11 开始，该关键字被弃用，因为标准委员会认为让用户直接与寄存器进行操作沟通是不安全的，且会干扰相关的编译器优化系统**

所以在 C++11 及以上的编译器中，该关键字将被**忽略**并抛出一个编译器警告信息，**不会导致编译过程停止**，只是编译器不会对该关键字进行处理

## bool 与 int 问题

**bool 不会与 int 有速度差距，bool 慢于 int 当且仅当使用 `std::vector` 作为容器时**

bool 根据编译器不同会在内部实现为 **unsigned char** 或 **char**，**在经过编译器的优化过程后产生的汇编代码与使用 int 没有本质区别（仅有 AT&T 指令字长后缀或 Intel 指令类型标识有区别）**

至于为什么 bool 在使用 `std::vector` 时会比 int 更慢，是因为不少编译器专门针对 `std::vector<bool>` 进行了特化处理，将其实现成了类似于 `std::bitset` 的情况，因此在此情况下各种涉及到**迭代器、随机访问和地址寻址**的操作就会涉及到位移处理和冗余内存分配，使得容器的效率急剧下降

<del>所以当你要用 `std::vector<bool>` 的时候为啥不用 `std::bitset`</del>
